enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,560 @@
+  choosing between, e.g. from `Warning` to `Critical`.
+- As an OpenShift user, I want to add labels to alerts to aide in routing and
+  triage.
+- As an OpenShift user, I want a supported way to add custom alerts using
+  metrics from components in the `openshift-*` namespaces.
+- As an OpenShift developer, I want the alerting rules shipped by my component
+  to be evaluated and forwarded to telemetry even when overridden by a user.
+
+### API Extensions
+
+We will be adding two new CRDs for alerting rules and relabeling:
+
+- `alertingrule.monitoring.openshift.io/v1alpha1`
+- `alertrelabelconfig.monitoring.openshift.io/v1alpha1`
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// AlertingRule represents a set of user-defined Prometheus rule groups containing
+// alerting rules -- recording rules are not allowed.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertingRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertingRule object.
+	Spec AlertingRuleSpec `json:"spec"`
+
+	// status describes the current state of this AlertOverrides object.
+	//
+	// +optional
+	Status AlertingRuleStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlertingRuleList is a list of AlertingRule objects.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertingRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertingRule objects.
+	Items []AlertingRule `json:"items"`
+}
+
+// AlertingRuleSpec is the desired state of an AlertingRule resource.
+//
+// +k8s:openapi-gen=true
+type AlertingRuleSpec struct {
+	// groups is a list of grouped alerting rules.
+	Groups []RuleGroup `json:"groups,omitempty"`
+}
+
+// RuleGroup is a list of sequentially evaluated alerting rules.
+//
+// +k8s:openapi-gen=true
+type RuleGroup struct {
+	// name is the name of the group.
+	Name string `json:"name"`
+
+	// interval is how often rules in the group are evaluated.  If not specified,
+	// it defaults to the global.evaluation_interval configured in Prometheus,
+	// which itself defaults to 1 minute.
+	//
+	// +optional
+	Interval string `json:"interval,omitempty"`
```

Comment:
This is a duration right? Have you considered using `metav1.Duration` instead of a string? This is what we would typically use in an API for a duration at least

It would be good to include details of the expected value here, for example:

```
// Expects an unsigned duration string of decimal numbers each with optional
// fraction and a unit suffix, eg "300ms", "1.5h" or "2h45m".
// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
```
And if this assumption is correct, you can use the following openapi validation
```
// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
// +kubebuilder:validation:Type:=string
```

Reply:
Yeah, that struck me as weird too, but the thing is, this entire `AlertingRule` type is a literal copy of the upstream `PrometheusRule` type from prometheus-operator. I'm using the same types they use, and the validation pattern is copy and pasted from the description in Prometheus. The purpose of the type is to give users an identical experience to using upstream `PrometheusRule` objects, except to disallow certain things like recording rules. So, it seems worthwhile to me to keep them the as similar as possible.

Reply:
Do you have a link to the upstream `PrometheusRule` type? It may be worth adding a link to this as a developer comment within the struct godoc so that future developers understand the relation between the two objects

I appreciate the sentiment of trying to keep it semantically similar to the upstream, however, in the past when I've had API reviews, I've been told that as long as you are API compatible (ie can convert between them), we should strive to give the best UX for our users independent of whether that means straying away from an upstream design. Repeating sins of the past arguments don't tend to fly in the API review sector in my experience.

If changing this to a duration either makes the user experience better, or makes validation/programming with the object easier, then IMO we should consider this change

Reply:
So, Prometheus actually has its own duration format:
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration

I've added the correct validations, and added a note that this is the Prometheus duration format, with a link to the upstream docs. I've also added a note at the top of the file that explains that the resource is basically a 1:1 copy of the upstream resource and why, along with a link to the upstream API docs.